File: internal/domain/displaystate.go
package domain

import (
	"fmt"
	"strings"
)

// GetDisplayState resolves the single status that earns a PR's ball.
// Worst state wins: CI fail > needs review > the rest.
func GetDisplayState(pr PullRequest) PrDisplayState {
	if pr.CI.State == CIFailure || pr.CI.Failed > 0 {
		return StateCIFailure
	}

	reviewBlocked := pr.ChangesRequested ||
		pr.ReviewDecision == "CHANGES_REQUESTED" ||
		(pr.Mergeable != nil && !*pr.Mergeable)
	if reviewBlocked {
		return StateReviewBlocked
	}

	if pr.Merged {
		return StateMerged
	}
	if pr.Draft {
		return StateDraft
	}
	if pr.MergeQueueState != "" && pr.MergeQueueState != "NONE" {
		return StateQueued
	}
	if pr.ReviewDecision == "APPROVED" {
		return StateReady
	}
	return StateOpen
}

var DisplayStateLabel = map[PrDisplayState]string{
	StateMerged:        "merged",
	StateDraft:         "draft",
	StateCIFailure:     "CI failing",
	StateReviewBlocked: "needs your review",
	StateQueued:        "merge queued",
	StateReady:         "approved",
	StateOpen:          "open / pending",
}

func DisplayStateDetail(pr PullRequest) string {
	state := GetDisplayState(pr)
	switch state {
	case StateMerged:
		return "already landed"
	case StateDraft:
		return "not requesting review"
	case StateReviewBlocked:
		if pr.Mergeable != nil && !*pr.Mergeable {
			return "merge conflict"
		}
		if pr.ChangesRequested {
			return "changes requested"
		}
	case StateCIFailure:
		failed := pr.CI.Failed
		if failed == 0 {
			failed = 1
		}
		return fmt.Sprintf("%d CI check failed", failed)
	case StateQueued:
		return humanizeQueueState(pr.MergeQueueState)
	case StateReady:
		return "checks passed · review complete"
	}

	if pr.CI.State == CIPending && pr.Approvals > 0 {
		return "checks running · review complete"
	}
	if pr.CI.State == CIPending {
		return "checks running · awaiting review"
	}
	if pr.CI.State == CISuccess && pr.Approvals > 0 {
		return "checks passed · awaiting merge"
	}
	if pr.CI.State == CISuccess {
		return "checks passed · awaiting review"
	}
	if pr.Approvals > 0 {
		return "review complete · awaiting checks"
	}
	return "awaiting checks and review"
}

func humanizeQueueState(queueState string) string {
	if queueState == "" || queueState == "NONE" || queueState == "QUEUED" {
		return "waiting in merge queue"
	}
	lowered := strings.ToLower(queueState)
	lowered = strings.ReplaceAll(lowered, "_", " ")
	return strings.TrimPrefix(lowered, "merge queue ")
}

func StateColorToken(state PrDisplayState) string {
	switch state {
	case StateCIFailure:
		return "ciFailure"
	case StateReviewBlocked:
		return "warning"
	case StateDraft:
		return "meta"
	case StateOpen:
		return "paper"
	default:
		return string(state)
	}
}

// BallGlyph is the one mark for that state. The active layer is a
// fisheye (◉) painted by the list — same ink, different shape. Review
// stays ◎. No logo rainbow.
func BallGlyph(state PrDisplayState) rune {
	switch state {
	case StateDraft:
		return '○'
	case StateReviewBlocked:
		return '◎'
	case StateQueued:
		return '◌'
	default:
		return '●'
	}
}

File: src/domain/displayState.ts
import type { PrDisplayState, PullRequest } from "./types.ts"

/**
 * Resolve the single status that earns a PR's ball color and inspector headline.
 * The order is intentionally product policy, not a rendering detail.
 */
export function getDisplayState(pr: PullRequest): PrDisplayState {
  if (pr.merged) return "merged"
  if (pr.draft) return "draft"
  if (pr.ci.state === "failure" || pr.ci.failed > 0) return "ci-failure"

  const reviewBlocked =
    pr.changesRequested ||
    pr.reviewDecision === "CHANGES_REQUESTED" ||
    pr.mergeable === false
  if (reviewBlocked) return "review-blocked"

  if (pr.mergeQueueState && pr.mergeQueueState !== "NONE") return "queued"
  if (pr.reviewDecision === "APPROVED" && pr.ci.state === "success") return "ready"
  return "open"
}

export const displayStateLabel: Record<PrDisplayState, string> = {
  merged: "merged",
  draft: "draft",
  "ci-failure": "CI failing",
  "review-blocked": "review / merge blocked",
  queued: "merge queued",
  ready: "ready to merge",
  open: "open / pending",
}

export function displayStateDetail(pr: PullRequest): string {
  const state = getDisplayState(pr)
  if (state === "merged") return "already landed"
  if (state === "draft") return "not requesting review"
  if (state === "review-blocked" && pr.mergeable === false) return "merge conflict"
  if (state === "review-blocked" && pr.changesRequested) return "changes requested"
  if (state === "ci-failure") return `${pr.ci.failed || 1} CI check failed`
  if (state === "queued") return humanizeQueueState(pr.mergeQueueState)
  if (state === "ready") return "checks passed · review complete"
  if (pr.ci.state === "pending" && pr.approvals > 0) return "checks running · review complete"
  if (pr.ci.state === "pending") return "checks running · awaiting review"
  if (pr.ci.state === "success" && pr.approvals > 0) return "checks passed · awaiting merge"
  if (pr.ci.state === "success") return "checks passed · awaiting review"
  if (pr.approvals > 0) return "review complete · awaiting checks"
  return "awaiting checks and review"
}

function humanizeQueueState(queueState: string | undefined): string {
  if (!queueState || queueState === "NONE" || queueState === "QUEUED") return "waiting in merge queue"
  return queueState
    .toLowerCase()
    .replaceAll("_", " ")
    .replace(/^merge queue /, "")
}
